README
libmatenotify is a fork of libnotify


NEWS
NEW in 0.5.2:
==============

Note that the 0.5.1 release was released from the 0.6 branch in error.
Sorry about that.

- Fixed #623096, Use correct variable in NULL check
- Added a macro NOTIFY_CHECK_VERSION

NEW in 0.5.0:
==============
- Support for version 1.1 of notification spec (Aurélien Gâteau, A. Walton)
- Fixed #608089, Applications aborted at finalizing when the dbus session is being closed. (William Jon McCann)
- Fixed #608082, Arabic text is misplaced in bubbles (RTL) (William Jon McCann)
- Fixed #608086, wrong return value of notify_init() (William Jon McCann)
- Fixed #612720, Does not compile with -DGSEAL_ENABLE (William Jon McCann)
- Fixed #613999, Use g_strcompress in notify-send to parse \n from command line (William Jon McCann)
- Fixed #610880, lazily initialize the daemon (Martin Pitt)


version 0.4.5 (20-November-2008):
	* Added GObject property accessors for notification IDs. Patch by Lucas
	  Rocha.
	* Added support for sending the closed reason in the "closed" signal
	  handler. (Ticket #139)
	* Added support for sending the status icon XID to the notification
	  daemon. This will allow better position tracking for context
	  notifications. Patch by Colin Walters.
	* Fixed a breakage where we were including gtkversion.h directly.

version 0.4.4 (27-February-2007):
	* Fixed a bug where a notification's ID could be reset when a different
	  notification was closed. Patch by jylefort. (Bug #94)
	* Fixed a crash when the D-BUS proxy was not being freed on notify_uninit,
	  which was problematic when used in a loadable module. (Bug #92)
	* Fixed a crash when a signal handler for the notification's closed signal
	  caused the notification to be destroyed. (Bug #116)
	* Fixed memory leaks when creating notifications. (Bug #112)
	* Fixed potential memory leaks where the function passed to
	  notify_notification_add_action to free the user data was not being
	  called. (Bug #119)

version 0.4.3 (8-October-2006):
	* Fixed the raw image data support to send the correct array type.
	  Patch by mderezynski. (Bug #68)
	* Fixed configure.ac to check for the required version of glib (2.6).
	  Patch by compnerd. (Bug #69)
	* Fixed invalid notification IDs when notify_notification_show() is called
	  on a notification that had already been closed or had expired. Now,
	  the notification will receive a new ID. Patch by Ed Catmur. (Bug #78)
	* Fixed the property change notification used when setting a GtkStatusIocn.
	  It was emitting "attach-icon" but should have been emitted "status-icon".
	  (Bug #81)
	* Fixed an invalid assertion in notify_notification_new_with_status_icon.
	  Patch by lucasr. (Bug #85)
	* notify_uninit is no longer called automatically at exit. Applications
	  and libraries should do this manually. Patch by berndth. (Bug #86)

version 0.4.2 (17-June-2006):
	* Add G_BEGIN_DECLS and G_END_DECLS to notify.h, so that libmatenotify can
	  be used in C++ applications properly.

version 0.4.1 (17-June-2006):
	* Added support for attaching to a GtkStatusIcon. Patch by chpe. (Bug #60)
	* Added GObject properties to NotifyNotification. Patch by chpe. (Bug #60)
	* Added API documentation, accessible through devhelp.
	* Fixed up libmatenotify.pc to support dependencies correctly. (Bug #58)
	* Fixed notify_uninit() to properly set _initted to FALSE after being
	  called. (Bug #56)

version 0.4.0 (26-April-2006):
	* Patch by M.S. to switch notify-send to use GOption instead of popt,
	  to add -v, --version, -h, and --hint options, and to rename
	  -T, --type to -c, --category. This also fixes assertions when calling
	  notify-send "". (Bug #41)
	* Fixed the short flag for --type to be -T in notify-send. (Bug #18)
	* Fixed the help in notify-send to indicate that --expire-timeout
	  requires milliseconds, not seconds. (Bug #28)
	* Fixed some C99 usage. Patch by jensgr. (Bug #44)
	* Fixed notify-send on PPC.

version 0.3.2 (23-January-2006):
	* Added back notify_get_server_info() and notify_get_server_caps().
	* Fixed to work with D-BUS 0.36 and higher. However, it's best to note
	  that due to bugs in versions of the D-BUS GLib bindings before 0.60,
	  raw image data cannot be sent in notifications. A warning will be
	  outputted to the console.
	* Added assertions to all of the API functions so that invalid data passed
	  won't crash the calling program.
	* Added automatic notification removal for non-expiring notifications and
	  notifications with actions when the calling application closes.
	* Fixed issues where the header files couldn't always be included in
	  some programs.
	* Fixed a lot of compiler errors and warnings.
	* Fixed C99 usage.
	* Added the ability to set user data on action callbacks on a per-action
	  basis.
	* Fixed notify-send to work without crashing.
	* Fixed the library to send messages appropriate to the D-BUS
	  specification.
	* Cleaned up the API a lot.
	* Renamed NOTIFY_TIMEOUT_* to NOTIFY_EXPIRES_*.
	* Removed notify_notification_show_and_forget().
	* Internal changes to remove the number of DBusGProxy connections, and
	  make sure it's initialized before using any D-BUS GValue-related code.
	  This fixes raw image sending.

version 0.3.0:
	* First release of the new GObject based API and simpler wire protocol

version 0.2.3:
	* Fix notify-send to work with icon parameters (bug #4308)

version 0.2.2 (10-August-2005):
	* Fixed many run-time issues with PPC and AMD64 (Martin Pitt)

version 0.2.1 (28-July-2005):
	* Fixed hint support for D-BUS 0.3x.

version 0.2.0 (28-July-2005):
	* Initial public release.


libmatenotify/notify.h
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 *
 * Copyright (C) 2004-2006 Christian Hammond
 * Copyright (C) 2010 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307, USA.
 */

#ifndef _LIBMATENOTIFY_NOTIFY_H_
#define _LIBMATENOTIFY_NOTIFY_H_

#include <glib.h>
#include <time.h>

#include <libmatenotify/notification.h>
#include <libmatenotify/notify-enum-types.h>
#include <libmatenotify/notify-features.h>

#ifdef __cplusplus
extern "C" {
#endif

/**
 * Initializes the notifications library.
 *
 * @param app_name The application name.
 *
 * @return TRUE if the library initialized properly.
 */
gboolean        notify_init (const char *app_name);

/**
 * Uninitializes the notifications library.
 *
 * This will be automatically called on exit unless previously called.
 */
void            notify_uninit (void);

/**
 * Returns whether or not the notification library is initialized.
 *
 * @return TRUE if the library is initialized, or FALSE.
 */
gboolean        notify_is_initted (void);

/**
 * Returns the name of the application set when notify_init() was called.
 *
 * @return The name of the application.
 */
const gchar    *notify_get_app_name (void);

/**
 * Returns the capabilities of the notification server.
 *
 * @return A list of capability strings. These strings must be freed.
 */
GList          *notify_get_server_caps (void);

/**
 * Returns the server notification information.
 *
 * The strings returned must be freed.
 *
 * @param ret_name         The returned product name of the server.
 * @param ret_vendor       The returned vendor.
 * @param ret_version      The returned server version.
 * @param ret_spec_version The returned specification version supported.
 *
 * @return TRUE if the call succeeded, or FALSE if there were errors.
 */
gboolean        notify_get_server_info (char **ret_name,
                                        char **ret_vendor,
                                        char **ret_version,
                                        char **ret_spec_version);

#ifdef __cplusplus
}
#endif

#endif /* _LIBMATENOTIFY_NOTIFY_H_ */


libmatenotify/notify.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 *
 * Copyright (C) 2004-2006 Christian Hammond <chipx86@chipx86.com>
 * Copyright (C) 2004-2006 Mike Hearn <mike@navi.cx>
 * Copyright (C) 2010 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307, USA.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>

#include "notify.h"
#include "internal.h"
#include "notify-marshal.h"

static gboolean         _initted = FALSE;
static char            *_app_name = NULL;
static DBusGProxy      *_proxy = NULL;
static DBusGConnection *_dbus_gconn = NULL;
static GList           *_active_notifications = NULL;
static int              _spec_version_major = 0;
static int              _spec_version_minor = 0;

gboolean
_notify_check_spec_version (int major,
                            int minor)
{
       if (_spec_version_major > major)
               return TRUE;
       if (_spec_version_major < major)
               return FALSE;
       return _spec_version_minor >= minor;
}

static gboolean
_notify_update_spec_version (void)
{
       char *spec_version;

       if (!notify_get_server_info (NULL, NULL, NULL, &spec_version)) {
               return FALSE;
       }

       sscanf (spec_version,
               "%d.%d",
               &_spec_version_major,
               &_spec_version_minor);

       g_free (spec_version);

       return TRUE;
}

/**
 * notify_init:
 * @app_name: The name of the application initializing libmatenotify.
 *
 * Initialized libmatenotify. This must be called before any other functions.
 *
 * Returns: %TRUE if successful, or %FALSE on error.
 */
gboolean
notify_init (const char *app_name)
{
        g_return_val_if_fail (app_name != NULL, FALSE);
        g_return_val_if_fail (*app_name != '\0', FALSE);

        if (_initted)
                return TRUE;

        _app_name = g_strdup (app_name);

        g_type_init ();

        _initted = TRUE;

        return TRUE;
}

/**
 * notify_get_app_name:
 *
 * Gets the application name registered.
 *
 * Returns: The registered application name, passed to notify_init().
 */
const char *
notify_get_app_name (void)
{
        return _app_name;
}

/**
 * notify_uninit:
 *
 * Uninitialized libmatenotify.
 *
 * This should be called when the program no longer needs libmatenotify for
 * the rest of its lifecycle, typically just before exitting.
 */
void
notify_uninit (void)
{
        GList *l;

        if (!_initted) {
                return;
        }

        if (_app_name != NULL) {
                g_free (_app_name);
                _app_name = NULL;
        }

        for (l = _active_notifications; l != NULL; l = l->next) {
                NotifyNotification *n = NOTIFY_NOTIFICATION (l->data);

                if (_notify_notification_get_timeout (n) == 0 ||
                    _notify_notification_has_nondefault_actions (n)) {
                        notify_notification_close (n, NULL);
                }
        }

        if (_proxy != NULL) {
            g_object_unref (_proxy);
        }

        _initted = FALSE;
}

/**
 * notify_is_initted:
 *
 * Gets whether or not libmatenotify is initialized.
 *
 * Returns: %TRUE if libmatenotify is initialized, or %FALSE otherwise.
 */
gboolean
notify_is_initted (void)
{
        return _initted;
}

DBusGConnection *
_notify_get_dbus_g_conn (void)
{
        return _dbus_gconn;
}

static void
on_proxy_destroy (DBusGProxy *proxy,
                  gpointer    data)
{
        _proxy = NULL;
}

DBusGProxy *
_notify_get_g_proxy (void)
{
        GError          *error;
        DBusGConnection *bus;

        if (_proxy != NULL)
                return _proxy;

        /* lazily initialize D-Bus connection */
        error = NULL;
        bus = dbus_g_bus_get (DBUS_BUS_SESSION, &error);
        if (error != NULL) {
                g_error_free (error);
                return NULL;
        }

        _proxy = dbus_g_proxy_new_for_name (bus,
                                            NOTIFY_DBUS_NAME,
                                            NOTIFY_DBUS_CORE_OBJECT,
                                            NOTIFY_DBUS_CORE_INTERFACE);
        dbus_g_connection_unref (bus);

        g_signal_connect (_proxy,
                          "destroy",
                          G_CALLBACK (on_proxy_destroy),
                          NULL);

        dbus_g_object_register_marshaller (notify_marshal_VOID__UINT_UINT,
                                           G_TYPE_NONE,
                                           G_TYPE_UINT,
                                           G_TYPE_UINT,
                                           G_TYPE_INVALID);

        dbus_g_object_register_marshaller (notify_marshal_VOID__UINT_STRING,
                                           G_TYPE_NONE,
                                           G_TYPE_UINT,
                                           G_TYPE_STRING,
                                           G_TYPE_INVALID);

        dbus_g_proxy_add_signal (_proxy,
                                 "NotificationClosed",
                                 G_TYPE_UINT,
                                 G_TYPE_UINT,
                                 G_TYPE_INVALID);
        dbus_g_proxy_add_signal (_proxy,
                                 "ActionInvoked",
                                 G_TYPE_UINT,
                                 G_TYPE_STRING,
                                 G_TYPE_INVALID);

        if (!_notify_update_spec_version ()) {
               return NULL;
        }

        return _proxy;
}

/**
 * notify_get_server_caps:
 *
 * Queries the server for its capabilities and returns them in a #GList.
 *
 * Returns: A #GList of server capability strings.
 */
GList *
notify_get_server_caps (void)
{
        GError         *error;
        char          **caps;
        char          **cap;
        GList          *result;
        DBusGProxy     *proxy;

        caps = NULL;
        result = NULL;

        proxy = _notify_get_g_proxy ();
        if (proxy == NULL) {
                return NULL;
        }

        error = NULL;
        if (!dbus_g_proxy_call (proxy,
                                "GetCapabilities",
                                &error,
                                G_TYPE_INVALID,
                                G_TYPE_STRV,
                                &caps,
                                G_TYPE_INVALID)) {
                g_error_free (error);
                return NULL;
        }

        for (cap = caps; *cap != NULL; cap++) {
                result = g_list_append (result, g_strdup (*cap));
        }

        g_strfreev (caps);

        return result;
}

/**
 * notify_get_server_info:
 * @ret_name: The resulting server name.
 * @ret_vendor: The resulting server vendor.
 * @ret_version: The resulting server version.
 * @ret_spec_version: The resulting version of the specification the server is
 *                    compliant with.
 *
 * Queries the server for its information, specifically, the name, vendor,
 * server version, and the version of the notifications specification that it
 * is compliant with.
 *
 * Returns: %TRUE if successful, and the variables passed will be set. %FALSE
 *          on failure.
 */
gboolean
notify_get_server_info (char **ret_name,
                        char **ret_vendor,
                        char **ret_version,
                        char **ret_spec_version)
{
        GError         *error;
        DBusGProxy     *proxy;
        char           *name;
        char           *vendor;
        char           *version;
        char           *spec_version;

        proxy = _notify_get_g_proxy ();
        if (proxy == NULL) {
                return FALSE;
        }

        error = NULL;
        if (!dbus_g_proxy_call (proxy,
                                "GetServerInformation",
                                &error,
                                G_TYPE_INVALID,
                                G_TYPE_STRING, &name,
                                G_TYPE_STRING, &vendor,
                                G_TYPE_STRING, &version,
                                G_TYPE_STRING, &spec_version,
                                G_TYPE_INVALID)) {
                g_error_free (error);
                return FALSE;
        }

        if (ret_name != NULL) {
                *ret_name = name;
        } else {
                g_free (name);
        }

        if (ret_vendor != NULL) {
                *ret_vendor = vendor;
        } else {
                g_free (vendor);
        }

        if (ret_version != NULL) {
                *ret_version = version;
        } else {
                g_free (version);
        }

        if (ret_spec_version != NULL) {
                *ret_spec_version = spec_version;
        } else {
                g_free (spec_version);
        }

        return TRUE;
}

void
_notify_cache_add_notification (NotifyNotification *n)
{
        _active_notifications = g_list_prepend (_active_notifications, n);
}

void
_notify_cache_remove_notification (NotifyNotification *n)
{
        _active_notifications = g_list_remove (_active_notifications, n);
}
